File: algorithms/96_UniqueBinarySearchTrees.java
class Solution {
    public int numTrees(int n) {
        if (n == 0) { return 0; }
        
        int[] arr = new int[n];
        for (int i=0; i<n;++i) {
            arr[i] = 0;
        }
        return dfs(n, arr);
    }
    
    public int dfs(int n, int[] arr) {
        if (n == 0) {
            return 1;
        } else if (arr[n-1] != 0) {
            return arr[n-1];
        }
        int total = 0;
        for (int i=0; i<n; ++i) {
            int left_count = dfs(i, arr);
            int right_count = dfs(n-i-1, arr);
            total += left_count * right_count;
        }
        arr[n-1] = total;
        return total;
    }
}

File: algorithms/96_UniqueBinarySearchTrees.py
class Solution:
    def numTrees(self, n):
        if not n: return 0
        self.pre_calculated = [0 for _ in range(n)]
        return self.dfs(n)
    
    def dfs(self, n):
        if not n:
            return 1
        elif self.pre_calculated[n - 1]:
            return self.pre_calculated[n - 1]
        
        total = 0
        for i in range(n):
            left_count = self.dfs(i)
            right_count = self.dfs(n-1-i)
            total += left_count * right_count
        self.pre_calculated[n - 1] = total
        return total
        